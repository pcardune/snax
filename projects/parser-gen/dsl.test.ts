const grammar = `
TOKEN_A = "a"
TOKEN_B = "b"

Root =
  | [TOKEN_A]
  | [TOKEN_B Rule]

Rule = 
  | [TOKEN_A Rule "some string"]
  | []
`;
import {
  parser,
  lexer,
  compileLexerToTypescript,
  compileGrammarToTypescript,
} from './dsl';
describe('dsl', () => {
  const tokens = lexer.parse(grammar);
  const root = parser.parseTokensOrThrow(tokens);
  test('compileLexerToTypescript', () => {
    const ts = compileLexerToTypescript(root, '/importRoot')._unsafeUnwrap();
    expect(ts).toMatchSnapshot();
  });

  test('compileParserToTypescript', () => {
    const ts = compileGrammarToTypescript(root, '/importRoot')._unsafeUnwrap();
    expect(ts).toMatchSnapshot();
  });
});
