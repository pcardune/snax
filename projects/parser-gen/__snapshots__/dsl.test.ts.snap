// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dsl compileLexerToTypescript 1`] = `
"
import { OrderedMap } from '/importRoot/utils/data-structures/OrderedMap';
import { RegexParser } from '/importRoot/regex-compiler/parser';
import { charSeq } from '/importRoot/nfa-to-dfa/regex-nfa';
import { ConstNFA } from '/importRoot/nfa-to-dfa/nfa';

const re = (s: string) => RegexParser.parseOrThrow(s).nfa();

export enum Token {
  TOKEN_A=\\"a\\",
  TOKEN_B=\\"b\\",
  IMPLICIT_2=\\"some string\\",
}

const patterns: OrderedMap<Token, {nfa:ConstNFA, ignore:boolean}> = new OrderedMap([
  [Token.TOKEN_A, {nfa: charSeq(\\"a\\"), ignore: false}],
  [Token.TOKEN_B, {nfa: charSeq(\\"b\\"), ignore: false}],
  [Token.IMPLICIT_2, {nfa: charSeq(\\"some string\\"), ignore: false}],
])

import { PatternLexer } from '/importRoot/lexer-gen/recognizer';
export const lexer = new PatternLexer(patterns);
"
`;

exports[`dsl compileParserToTypescript 1`] = `
"

import { buildParser, ParseNode } from '/importRoot/grammar/top-down-parser';
import { GrammarSpec } from '/importRoot/grammar/grammar';
export enum Rules {
  Root = \\"Root\\",
  Rule = \\"Rule\\",
}

const grammarSpec: GrammarSpec<Rules|Token> = {
  [Rules.Root]:[
    [Token.TOKEN_A, ],
    [Token.TOKEN_B, Rules.Rule, ],
  ],
  [Rules.Rule]:[
    [Token.TOKEN_A, Rules.Rule, Token.IMPLICIT_2, ],
    [],
  ],
};
export const parser = buildParser(grammarSpec);"
`;
